2.x Add ImageHelper filters to customize paths

mcaskill opened the pull request:
Related:

- #2458
- #2453

## Issue

As reported in aforementioned issue and pull request, the `ImageHelper::theme_url_to_dir()` method is unable to resolve a URL/path in a non-standard theme directory structure.

This is the case if your "stylesheet" directory is located in a subdirectory of your theme `my-theme/theme` as was initially proposed in Timber's v2 starter-theme and employed by various theme boilerplates.

## Solution

The solution, as [proposed](https://github.com/timber/timber/pull/2458#issuecomment-1471396358) by @gchtr, is to add filters in `ImageHelper` to allow one to customize the logic of converting a URL to a path.

> For example, we could add a `timber/theme_url_to_dir` filter that allows you to filter the result of `Timber\ImageHelper::theme_url_to_dir()`.

I've added four filters:

- `ImageHelper::analyze_url()`
	- `timber/image_helper/pre_analyze_url` to short-circuit the method's default URL analysis logic.
	- `timber/image_helper/analyze_url` to mutate the default analyzed URL components.
- `ImageHelper::theme_url_to_dir()`
	- `timber/image_helper/pre_theme_url_to_dir` to short-circuit the method's default path resolution logic.
	- `timber/image_helper/theme_url_to_dir` to mutate the default resolved path.

I've added filters to `analyze_url()` since it's the one that calls `theme_url_to_dir()` and also contains opinionated logic for analyzing the URL.

Maybe the short-circuit filters could still call their non-pre filters, this would involve wrapping the default logic in a "did pre-filter return null".

Here's how its being used in a client project:

```php
public function register_filters() : void {
	add_filter( 'timber/image_helper/pre_theme_url_to_dir', [ $this, 'filter_timber_image_helper_pre_theme_url_to_dir' ], 10, 2 );
}

public function filter_timber_image_helper_pre_theme_url_to_dir( ?string $path, string $src ): ?string {
	if ( is_null( $path ) && str_starts_with( $src, $this->theme_uri ) ) {
		$path = str_replace( $this->theme_uri, '', strtok( $src, '?' ) );
		$path = $this->theme_path . '/' . ltrim( $path, '/' );
	}

	return $path;
}
```

## Impact

Adding an extra hooks could affect performance but it should be minor.

## Usage Changes

The current logic of those methods does not change, filters are added to allow one to handle them early or mutate the default resulting values.

## Considerations

See discussions from #2458

## Testing

- Modified `install-wp-tests.sh` to ensure `TMPDIR` is canonical.
- Fixed tests of `TestTimberImageHelperInternals::testAnalyzeURLTheme()`.
- Added tests for filters in `ImageHelper::analyze_url()`.
- Added tests for filters in `ImageHelper::theme_url_to_dir()`.

mcaskill committed:
Fix unit test comment in ImageHelper

mcaskill committed:
Move comments above related array element in ImageHelper

mcaskill committed:
Add filters to ImageHelper::analyze_url

Added:
- Filter "timber/image_helper/pre_analyze_url" to short-circuit the method's default URL analysis logic.
- Filter "timber/image_helper/analyze_url" to mutate the analyzed URL components.

mcaskill committed:
Add filters to ImageHelper:: theme_url_to_dir

Added:
- Filter "timber/image_helper/pre_theme_url_to_dir" to short-circuit the method's default path resolution logic.
- Filter "timber/image_helper/theme_url_to_dir" to mutate the resolved path.

mcaskill committed:
Rename filter parameter in ImageHelper::analyze_url

Renamed from `$components` to `$info`.

mcaskill committed:
Add types to ImageHelper methods analyze_url and theme_url_to_dir

mcaskill committed:
Add tests for filters in ImageHelper::analyze_url

mcaskill committed:
Decouple default logic from ImageHelper::analyze_url

Moved original URL parsing from `analyze_url()` to new method `_analyze_url()`.

Changed logic of filters; "timber/image_helper/pre_analyze_url" controls whether the original logic is called, to ensure the "timber/image_helper/analyze_url" filter is always applied.

mcaskill committed:
Decouple default logic from ImageHelper::theme_url_to_dir

Moved original URL conversion from `theme_url_to_dir()` to new method `_theme_url_to_dir()`.

Changed logic of filters; "timber/image_helper/pre_theme_url_to_dir" controls whether the original logic is called, to ensure the "timber/image_helper/theme_url_to_dir" filter is always applied.

mcaskill committed:
Update tests for filters in ImageHelper::analyze_url

mcaskill committed:
Ensure `TMPDIR` is canonical in tests

If any symbolic links in `TMPDIR` are not resolved, usage of `realpath()` risks paths being broken such as by `ImageHelper::analyze_url()` calling `URLHelper::remove_url_component()` to remove `WP_CONTENT_DIR`.

Given a WordPress installation at:

```sh
/var/folders/XX/XXXXXXXXXX/X/wordpress/
```

And the following URL:

```sh
http://example.org/wp-content/themes/timber-test-theme/assets/images/cardinals.jpg
```

The `ImageHelper::theme_url_to_dir()` method would return the following absolute path (which internally calls `realpath()`):

```sh
/private/var/folders/XX/XXXXXXXXXX/X/wordpress/wp-content/themes/timber-test-theme/assets/images/cardinals.jpg
```

The `ImageHelper::analyze_url()` method would return the following components:

```php
[
  'url'       => 'http://example.org/wp-content/themes/timber-test-theme/assets/images/cardinals.jpg',
  'absolute'  => true,
  'base'      => 2,
  'subdir'    => '/private/themes/timber-test-theme/assets/images',
  'filename'  => 'cardinals',
  'extension' => 'jpg',
  'basename'  => 'cardinals.jpg',
]
```

As illustrated by `subdir`, the absolute path is broken by `URLHelper::remove_url_component()` striping the symbolically linked path from the canonical path.

mcaskill committed:
Update TestTimberImageHelperInternals

Changed:
- Added `set_up()` and `tear_down()` from `TestExternalImage` to prepare test environment for content directory test.
- Fixed tests of `testAnalyzeURLTheme()` based on `TestExternalImage`.

These fixes will help with adding tests for `ImageHelper::theme_url_to_dir()`.

mcaskill committed:
Add tests for filters in ImageHelper::theme_url_to_dir

mcaskill committed:
Update TestTimberImageHelper

Changed:
- Added `set_up()` and `tear_down()` to match `TestTimberImageHelperInternals` (from `TestExternalImage`) to prepare test environment for content directory test.

gchtr commented:
Adding filters is definitely the way to go. Because this isn’t a breaking change and we want to focus on getting 2.0 ready, I switched this to `2.x Future`.

mcaskill committed:
Merge branch '2.x' into 2.x-add-image-helper-filters

* 2.x: (95 commits)
  Update statements to Weak implicit
  Check expected values as suggested by @nlemoine
  fix: phpstan issues & use fully qualified imports
  Refactor file models: update phpdoc
  Refactor file models: remove file property
  Refactor file models
  Fix coding style issue
  Update method descriptions
  Update return types and documentation
  Fix is_external() to work with domains only
  Fix test
  Improve URLHelper is_local method
  Improve docblocks of URLHelper methods is_local and is_external
  Replace strstr in URLHelper is_external method
  Fix shortened return in URLHelper is_local method
  remove error_log and prefix functions
  add check to image create functions
  Run coding standard checks only with highest or stable dependencies
  Add CS fixes to ignored revs
  Fix some CS issues
  ...

Moustachos commented:
@gchtr any update on this one now that 2.x is stable? :)

Thanks!

Levdbas commented:
Hi @Moustachos , we have this in mind for a first minor release for Timber 2. Release date is TBD.

Moustachos commented:
Thanks @Levdbas 👍

gchtr commented on src/ImageHelper.php:
```diff
@@ -516,26 +516,63 @@ public static function set_sideload_image_upload_dir(array $upload)
+        }
+
+        /**
+         * Filters the array of anlayzed URL components.
```
```suggestion
         * Filters the array of analyzed URL components.
```

gchtr commented on src/ImageHelper.php:
```diff
@@ -516,26 +516,63 @@ public static function set_sideload_image_upload_dir(array $upload)
+     * @param  string $url A URL (absolute or relative) pointing to an image.
+     * @return array<string, mixed> An array (see keys in code below).
+     */
+    private static function _analyze_url(string $url): array
```
According to [PSR-12](https://www.php-fig.org/psr/psr-12/), underscores should not be used for method names.

> Method names MUST NOT be prefixed with a single underscore to indicate protected or private visibility. That is, an underscore prefix explicitly has no meaning.

Yes, we still have some of those in the codebase, but for new private functions, we usually don’t introduce underscore prefixes.

This means that we should find a new name, which is the hard part here. Maybe we can use one of the following?

- `get_url_components()`
- `run_analyze_url()`

mcaskill replied:
Renamed `_analyze_url` to `get_url_components`.

gchtr commented on src/ImageHelper.php:
```diff
@@ -576,16 +614,52 @@ public static function analyze_url($url)
+     * @param string  $src A URL (http://example.org/wp-content/themes/twentysixteen/images/home.jpg).
+     * @return string Full path to the file in question.
+     */
+    private static function _theme_url_to_dir(string $src): string
```
Same as for `_analyze_url()`, I wouldn’t use an underscore prefix here.

```suggestion
    private static function get_dir_from_theme_url(string $src): string
```

mcaskill replied:
Renamed `_theme_url_to_dir` to `get_dir_from_theme_url`.

gchtr reviewed (changes requested):
Thanks for taking this on, @mcaskill. And sorry for the longer wait.

I really appreciate that you took the time to carefully craft the DocBlocks for the filters with proper descriptions 🙌.

I just have on issue with the underscore prefixes in the private function names. Otherwise, this looks good to me and I think we can merge this for the next minor release.

mcaskill committed:
Fix typo in image helper filter block comment

mcaskill committed:
Rename private methods in ImageHelper

Changed:
- Renamed `_analyze_url` to `get_url_components`.
- Renamed `_theme_url_to_dir` to `get_dir_from_theme_url`.

gchtr commented:
This looks good from my side 👍.

@nlemoine Can you have a quick look and merge this, if it looks alright to you, too?

nlemoine reviewed (approved):
Still feels a bit odd to handle such things but doesn't hurt 👍
